tests/connections.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import sys
+import os
+import infra.proc
+
+import e2e_args
+import getpass
+import os
+import time
+import logging
+import multiprocessing
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.jsonrpc
+import json
+import contextlib
+import resource
+import psutil
+
+from loguru import logger as LOG
+
+
+def run(args):
+    hosts = ["localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        primary_pid = primary.remote.remote.proc.pid
+        max_fds = 50
+        num_fds = psutil.Process(primary_pid).num_fds()
```

Comment:
Could this ever be over or close to 50?

Reply:
We could modify CCF to open an arbitrary amount of additional fds in the future. It doesn't seem particularly likely, do you have something specific in mind?

If you're concerned about this, we could make max_fds = num_fds + 50

Reply:
It's just that below you do: (max_fds - num_fds)*2 and that is how many connections you are openning, so if num_fds is close to 50 then they won't be that many

Reply:
No but it’s enough to go past max_fds. I’ll change it to an offset, that way even if num_fds goes up dramatically in the future for some reason, we’re still opening way more connections than we have free fds.